src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java

```diff
@@ -29,7 +42,75 @@ public Path getBasePath() {
 
   @NotNull
   public Path getCourseFilePath() {
-    return getBasePath().resolve(Paths.get(Project.DIRECTORY_STORE_FOLDER, "a-plus-project"));
+    return getBasePath().resolve(Paths.get(Project.DIRECTORY_STORE_FOLDER, "a-plus-project.json"));
+  }
+
+  /**
+   * Creates a local course file containing the given URL. If a course file already exists (even if
+   * it was created with a different URL), this method does nothing and returns false.
+   * @param sourceUrl The URL that is added to the course file.
+   * @return True if the course file was created successfully, false if a course file already
+   *         exists.
+   * @throws IOException If an IO error occurs while creating the file.
+   */
+  public boolean createCourseFile(@NotNull URL sourceUrl) throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    if (courseFile.isFile()) {
+      return false;
+    }
+
+    JSONObject jsonObject = new JSONObject();
+    jsonObject.put("url", sourceUrl.toString());
+    FileUtils.writeStringToFile(courseFile, jsonObject.toString(), StandardCharsets.UTF_8);
+    return true;
+  }
+
+  /**
+   * Parses the stored URL from the course file and returns it, or null if the course file doesn't
+   * exist.
+   * @return The URL in the course file, or null if the course file doesn't exist.
+   * @throws IOException   If an IO error occurs while reading the course file.
+   * @throws JSONException If the JSON in the course file is malformed.
+   */
+  @Nullable
+  public URL getCourseFileUrl() throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    if (!courseFile.isFile()) {
+      return null;
+    }
+
+    // TODO: the course file is written as UTF-8, ensure that this always works
+    JSONTokener tokenizer = new JSONTokener(new FileInputStream(courseFile));
+    JSONObject jsonObject = new JSONObject(tokenizer);
+    return new URL(jsonObject.getString("url"));
+  }
+
+  private static final Object courseFileLock = new Object();
+  
+  /**
+   * Adds an entry for the given module to the course file. If an entry exists with the same name,
+   * it is overwritten.
+   * @param module The module for which an entry is added.
+   * @throws IOException   If an IO error occurs (for an example, the course file doesn't exist).
+   * @throws JSONException If the existing JSON in the course file is malformed.
+   */
+  public void addCourseFileEntry(@NotNull Module module) throws IOException {
+    File courseFile = getCourseFilePath().toFile();
+    synchronized (courseFileLock) {
```

Comment:
this is quite a long `syncronized` block... do you think some parts could be extracted? for example, is there a chance, that something will write the `courseFile` config file while this block is being executed :question: